newIDE/app/src/UI/Search/UseSearchStructuredItem.js

```diff
@@ -0,0 +1,187 @@
+  chosenCategory: ?ChosenCategory,
+  chosenFilters: Set<string>
+): ?Array<SearchItem> => {
+  const searchApiRef = React.useRef<?any>(null);
+  const [searchResults, setSearchResults] = React.useState<?Array<SearchItem>>(
+    null
+  );
+
+  // Keep in memory a list of all the items, shuffled for
+  // easing random discovery of items when no search is done.
+  const shuffledSearchItems: ?Array<SearchItem> = React.useMemo(
+    () => {
+      if (!searchItemsById) return null;
+
+      return shuffle(
+        Object.keys(searchItemsById).map(id => searchItemsById[id])
+      );
+    },
+    [searchItemsById]
+  );
+
+  // Index items that have been loaded.
+  React.useEffect(
+    () => {
+      if (!searchItemsById) {
+        // Nothing to index - yet.
+        return;
+      }
+
+      const startTime = performance.now();
+      if (searchApiRef.current) {
+        searchApiRef.current = null;
+      }
+
+      try {
+        const newSearchApi = new Fuse(Object.values(searchItemsById), {
+          keys: [
+            { name: 'name', weight: 2 },
+            { name: 'shortDescription', weight: 1 },
+          ],
+          minMatchCharLength: 3,
+          threshold: 0.35,
+          includeMatches: true,
+          ignoreLocation: true,
+          ignoreFieldNorm: true,
+        });
+
+        const totalTime = performance.now() - startTime;
+        console.info(
+          `Indexed ${
+            Object.keys(searchItemsById).length
+          } items in ${totalTime.toFixed(3)}ms.`
+        );
+        searchApiRef.current = newSearchApi;
+      } catch (error) {
+        console.error('Error while indexing items: ', error);
+      }
+    },
+    [searchItemsById]
+  );
+
+  // Update the search results according to the items, search term
+  // chosen category and chosen filters.
+  const searchApi = searchApiRef.current;
+  React.useEffect(
+    () => {
+      let discardSearch = false;
+      if (!searchText) {
+        setSearchResults(
+          filterSearchItems(shuffledSearchItems, chosenCategory, chosenFilters)
+        );
+      } else {
+        if (!searchItemsById || !searchApi) {
+          console.info(
+            'Search for items skipped because items are not ready - will be retried when ready'
+          );
+          return;
+        }
+
+        const startTime = performance.now();
+        const results = searchApi.search(`'${searchText}`);
+        const totalTime = performance.now() - startTime;
+        console.info(
+          `Found ${results.length} items in ${totalTime.toFixed(3)}ms.`
+        );
+        if (discardSearch) {
+          console.info(
+            'Discarding search results as a new search was launched.'
+          );
+          return;
+        }
+        setSearchResults(
+          filterSearchItems(
+            results
+              // .filter(result => result.score <= 0.3)
+              .map(result => ({
+                ...result.item,
+                matches: result.matches,
```

Comment:
I wonder if we should rather return as results:
`?Array<{|item: SearchItem, matches: Array<{| key: string, indices: number[][] |}> |}>`

To avoid "polluting the item with matches".